util/autogen_banner.py

```diff
@@ -0,0 +1,49 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from textwrap import wrap
+
+LICENSE_BANNER = [
+    "Copyright lowRISC contributors.",
+    "Licensed under the Apache License, Version 2.0, see LICENSE for details.",
+    "SPDX-License-Identifier: Apache-2.0",
+]
+
+AUTOGEN_BANNER = [
+    "-------------------------------------------------------------------------",
+    "-------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! --------",
+    "DO NOT EDIT THIS FILE DIRECTLY. GENERATED WITH THE FOLLOWING COMMAND:",
+    "-------------------------------------------------------------------------",
+]
+
+
+def get_autogen_banner(command: str, comment=None) -> str:
+    """Returns a commented out auto-generated code warning banner.
+
+    command is a fully formatted string representing what command was used
+    to auto-generate the source.
+    comment is the style of comment supported by the file type.
+    """
+    command_lines = wrap(command.strip(), width=len(AUTOGEN_BANNER[0]))
```

Comment:
You need to configure this a bit more carefully, since textwrap is not actually intended for code. You want:
```python
command_lines = wrap(
  command,
  width=MAX_LEN,
  drop_whitespace=True,
  break_long_words=False,
  break_on_hyphens=False,
  initial_indent='  '
)
```
You then want to include backslashes like Michael requested:
```python
for i in range(0, len(command_lines)-1):
  command_lines += ' \\'
```

You will want `MAX_LEN` to be its own constant.

Reply:
Adding backslash breaks the CI. GCC does not like it:
```
FAILED: sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o 
/opt/buildcache/riscv/bin/riscv32-unknown-elf-gcc -Isw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta  -I../../s/sw/device/tests/autogen -I/home/vsts/work/1/s -I/home/vsts/work/1/a/build-out -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Werror -std=c11 -g -nostdinc -isystem/home/vsts/work/1/s/sw/device/lib/base/freestanding -fno-asynchronous-unwind-tables -fno-common -ffunction-sections -Wimplicit-fallthrough -Wswitch-default -Wno-error=unused-function -Wno-unused-parameter -Wno-sign-compare -Wno-missing-field-initializers -Os -fvisibility=hidden -Wa,--no-pad-sections -Wstrict-prototypes -march=rv32imc -mabi=ilp32 -mcmodel=medany -MD -MQ 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o' -MF 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o.d' -o 'sw/device/tests/autogen/971261f@@plic_all_irqs_test_lib@sta/plic_all_irqs_test.c.o' -c ../../s/sw/device/tests/autogen/plic_all_irqs_test.c
../../s/sw/device/tests/autogen/plic_all_irqs_test.c:8:1: error: multi-line comment [-Werror=comment]
    8 | // util/topgen.py -t hw/top_earlgrey/data/top_earlgrey.hjson \
      | ^
cc1: all warnings being treated as errors
```

Reply:
It should work if you add another trailing space (because then the C code is essentially "escaping" a space, rather than the newline). It's a bit ugly, though :-/

Reply:
It doesn't seem worth the trouble - lets just skip it?

Reply:
yeah let's just skip it then.

Reply:
backslashes have been problematic in autogen comments. I don't see consistent results from auto formatting. Lets leave that out unless there is a good way to handle them. The config options proposed by @mcy seem reasonable to me.

Reply:
Done.